### PR TITLE
Drop EOL Python, simplify PyPy situation, improve testing

### LINKED ISSUE #70: incompatible requirements
https://github.com/scipy/oldest-supported-numpy/blob/b0b3376f6838abd91a9988b936f70939d4acb775/setup.cfg#L45 and https://github.com/scipy/oldest-supported-numpy/blob/b0b3376f6838abd91a9988b936f70939d4acb775/setup.cfg#L70

would both apply for users of pypy on macos, but require different versions of numpy.  There may be other examples.

### LINKED ISSUE #72: Drop support for EOL pythons?
I think we should drop the markers for 3.5 and probably 3.6, and just raise `python_requires` appropriately. Those on old python versions will still get functioning `oldest-supported-numpy` packages, and almost by definition, those old language versions will not receive any new architecture support. However, they still have a maintenance cost, as seen in cases like #71, so I think we should drop them.

Python 3.7 is also [EOL](https://peps.python.org/pep-0537/) this month, but it's a borderline case.

### PR DESCRIPTION
**Pull request by h-vetinari**
Addressing a couple of inconsistencies & clean-ups:
* we are mixing availability of wheels with compatibility (inconsistently; e.g. there are no wheels for s390x or loongarch64)
* we occasionally produce more than one pin for invalid platform combinations
* we are carrying around cruft from python versions that are long dead

What I'm doing:
* drop old Python
* reflect compatibility not just on availability of wheels (PyPy, ppc, etc.)
* substantially expand testing, becoming strict on all valid platform combinations in the expectation of zero or one pins
* remove all skips to also test invalid platform combinations (so that we test at least there's maximum 1 pin)
* the test module become a much more easily readable documentation of what setup.cfg is supposed to reflect
* reflect the existence of win-arm64 & linux-ppc64le

What should still be improved:
* we should make it clear on which architectures AIX/OS400 is supported. I'm now using `"x86", "x86_64", "s390x"`, which is broadly equivalent to what it was set to before (though I removed `arm64` from that list, because it's +/- equivalent to `aarch64` that was already marked as invalid)

Closes #70
Closes #71
Closes #72

### CODE REVIEW

File: tests/test_dependencies.py [hunk 1]
```diff
@@ -69,20 +75,43 @@ def test_has_at_most_one_pinned_dependency(
...
+            expect_pin = True  # as of 1.19.2
+        case ["Darwin", "arm64"]:
+            expect_pin = True  # as of 1.21
+        # no official wheels, but register minimum compatibility
+        case ["Linux", "ppc64le"]:
+            expect_pin = True  # support +/- equivalent to Linux/aarch64, but no wheels
+        case ["Linux", "s390x"]:
+            expect_pin = True  # as of 1.17.5
+        case ["Linux", "loongarch64"]:
+            expect_pin = True  # as of 1.22
```
Comment (h-vetinari) [inline tests/test_dependencies.py:97]: I find this a very compact and readable crystallization of what we're trying to achieve in the setup.cfg, in addition to the distinction of whether wheels are available or not (we could add more testing for this if we whip up some emulation, see #73).

File: tests/test_dependencies.py [hunk 2]
```diff
@@ -69,20 +75,43 @@ def test_has_at_most_one_pinned_dependency(
...
+        case ["AIX", _]:
+            expect_pin = True  # as of 1.16
+        case ["OS400", _]:
+            # only supported on CPython 3.9, see above
+            expect_pin = True  # as of 1.23.3
```
Comment (h-vetinari) [inline tests/test_dependencies.py:102]: Same here as above: realistically, our compatibility resp. pins always come in pairs of platform/architecture, so having something just based on `platform_system` seems wrong to me.

**Review by h-vetinari — COMMENTED** [review]
Some notes for ease of reviewing

File: setup.cfg [hunk 3]
```diff
@@ -54,15 +49,14 @@ install_requires =
     numpy==1.22.2; platform_machine=='loongarch64' and python_version<'3.11'
```
Comment (radoering) [inline setup.cfg:49]:
```suggestion
    numpy==1.22.2; platform_machine=='loongarch64' and python_version>='3.8' and python_version<'3.11'
```

numpy 1.22.2 does not support Python 3.7.

File: setup.cfg [hunk 4]
```diff
@@ -16,33 +16,28 @@ version = 2022.11.19
 #            versions that are no longer supported. This may break installing
 #            older versions of packages that depend on this meta-package!
 [options]
-python_requires = >=3.5
+python_requires = >=3.7
```
Comment (rgommers) [inline setup.cfg:19]: So you think the `Important: ...` comment right above is incorrect? I think that's indeed the case (as long as `python_requires` gets bumped, users of older Python versions should get an older version of this meta-package), so let's edit that comment.

File: setup.cfg [hunk 5]
```diff
@@ -16,33 +16,28 @@ version = 2022.11.19
...
 
     # numpy 1.19 was the first minor release to provide aarch64 wheels, but
-    # wheels require fixes contained in numpy 1.19.2 (For Python 3.5: support
-    # was dropped in 1.19 so numpy 1.18.5 can be built from source instead).
-    numpy==1.18.5; python_version=='3.5' and platform_machine=='aarch64' and platform_python_implementation != 'PyPy'
-    numpy==1.19.2; python_version=='3.6' and platform_machine=='aarch64' and platform_python_implementation != 'PyPy'
-    numpy==1.19.2; python_version=='3.7' and platform_machine=='aarch64' and platform_python_implementation != 'PyPy'
-    numpy==1.19.2; python_version=='3.8' and platform_machine=='aarch64' and platform_python_implementation != 'PyPy'
+    # wheels require fixes contained in numpy 1.19.2
+    numpy==1.19.2; python_version=='3.7' and platform_machine in 'aarch64|ppc64le' and platform_system!='AIX' and platform_python_implementation != 'PyPy'
```
Comment (rgommers) [inline setup.cfg:34]: Is the comment above also reflective of the situation on `ppc64le`?

File: setup.cfg [hunk 6]
```diff
@@ -54,15 +49,14 @@ install_requires =
...
-    numpy==1.19.3; python_version=='3.9' and platform_system not in 'OS400' and platform_machine not in 'arm64|loongarch64' and platform_python_implementation != 'PyPy'
+    numpy==1.14.5; python_version=='3.7' and platform_machine not in 'arm64|aarch64|ppc64le|loongarch64' and platform_system!='AIX' and platform_python_implementation != 'PyPy'
+    numpy==1.17.3; python_version=='3.8' and platform_machine not in 'arm64|aarch64|ppc64le|s390x|loongarch64' and platform_python_implementation != 'PyPy'
+    numpy==1.19.3; python_version=='3.9' and platform_system not in 'OS400' and platform_machine not in 'arm64|loongarch64'
     # Note that 1.21.3 was the first version with a complete set of 3.10 wheels,
     # however macOS was broken and it's safe to build against 1.21.6 on all platforms (see gh-28 and gh-45)
-    numpy==1.21.6; python_version=='3.10' and platform_machine!='loongarch64' and platform_python_implementation != 'PyPy'
-    numpy==1.23.2; python_version=='3.11' and platform_python_implementation != 'PyPy'
+    # we want to exclude win-arm64, but cannot write "not (...)" in this configuration format, thus use De Morgan's law
```
Comment (rgommers) [inline setup.cfg:57]: I suggest deleting this comment or rephrasing to a simple "the last condition excludes win-arm64". I wasn't familiar with that law, and it's more confusing that helpful to reference it.

Comment (rgommers) [reply setup.cfg:57]: More importantly: these are the default requirements, and I don't see win-arm64 for py310/py311 special-cased elsewhere. So this seems wrong. Or am I missing the point of this change?

**Review by rgommers — COMMENTED** [review]
Thanks @h-vetinari. This looks quite good overall. The testing changes seem helpful; I didn't review them in a lot of detail, but they're quite understandable and I'll do a bit more checking before merging.

File: setup.cfg (hunk 5 above)
Comment (h-vetinari) [reply setup.cfg:34]: If we're talking compatibility only, ppc was added (with tests) in conda-forge for 1.16 and passed.

I guess realistically aarch/ppc should have _better_ compatibility than s390x, where we're only claiming 1.17.5.

That said, I just wanted to add it at all, and aarch seemed like a good reference. I'm happy to put whatever between 1.16 & 1.19 as a baseline.

File: setup.cfg (hunk 6 above)
Comment (h-vetinari) [reply setup.cfg:57]: In all the previous conditions, win-arm64 is excluded by virtue of `arm64` being excluded. Now that osx-arm64 is not exceptional anymore, we need to distinguish the windows-case specifically.

> I suggest deleting this comment or rephrasing to a simple "the last condition excludes win-arm64". I wasn't familiar with that law, and it's more confusing that helpful to reference it.

I guess my math studies show a bit 😋
The rule is simple enough (`not (A & B) == (not A) or (not B)`), and I just gave it its "proper" name. I left a comment because I find the condition unintuitive to read, so I though it's better to note why it came about. Happy to adapt or remove it though.

File: setup.cfg (hunk 4 above)
Comment (h-vetinari) [reply setup.cfg:19]: Sorry for overlooking that comment. I haven't verified it, but it seems obvious to me that someone on py<3.7 would be even able to install a package that says `python_requires >= 3.7`.

I just thought it'd be worth to stop carrying around cruft for very old python versions anymore (we could even remove 3.7, it's EOL as of this month), but if you want to be sure, I can re-add them.

Comment (rgommers) [reply setup.cfg:19]: > can re-add them.

No need to re-add, they're already included. The condition is `>= 3.7`, so 3.7 is there.

File: setup.cfg (hunk 5 above)
Comment (rgommers) [reply setup.cfg:34]: `aarch64` is set to 1.19.2 here because 1.19 is the first minor version that added wheels for that platform (and the `.2` because of a bug fix): https://pypi.org/project/numpy/1.19.2/#files.

There are no wheels for `ppc64le`, so I don't know why this requires special-casing here, or a change at all.

> That said, I just wanted to add it at all,

It is added, _every_ platform is. It's covered by the default requirements unless it's special-cased for a specific reason (and if so, it's then excluded from the relevant default requirement). So I suggest you revert this change.

File: setup.cfg (hunk 6 above)
Comment (rgommers) [reply setup.cfg:57]: > In all the previous conditions, win-arm64 is excluded by virtue of `arm64` being excluded.

That's an accident, because the code is specifically written for macOS arm64 - which was the only platform to use `arm64` as an identifier. The special cases are for:
```
platform_machine=='arm64' and platform_system=='Darwin'
```
which is a single platform. So the intent in the default requirements is to exclude that same macOS arm64 condition. That is the right fix here. The current change for Python 3.10/3.11 must be wrong, you cannot exclude a platform from the defaults without including it elsewhere as a special case.

Comment (h-vetinari) [reply setup.cfg:57]: > you cannot exclude a platform from the defaults without including it elsewhere as a special case.

I maintain that this is the correct approach if that platform doesn't have _any_ compatibe numpy version yet (that I'm aware of) - i.e. it falls back to plain `numpy`, and that's intentional.

Unless you tell me that someone has compiled and tested 1.25 (or whatever) for win-arm64 successfully, then we can set that.

File: setup.cfg (hunk 5 above)
Comment (rgommers) [reply setup.cfg:34]: Those were added each time in response to concrete bug fixes that landed in a particular NumPy release. The rule is: every platform is covered by the default requirements, unless there's a reason to bump up the min version. I'd like to stick to that, since it has worked well enough, and any other change has a nontrivial change of breaking something.

File: tests/test_dependencies.py [hunk 7]
```diff
@@ -1,56 +1,62 @@
...
+        case ["Linux", _]:
+            valid = True   # otherwise, Linux is everywhere
+        case ["Darwin", ("x86_64" | "arm64")]:
+            valid = True
+        case ["Windows", ("x86" | "x86_64" | "arm64")]:
+            valid = True
+        case [("AIX" | "OS400"), ("arm64" | "aarch64" | "ppc64le" | "loongarch64")]:
+            valid = False  # express "everything but arm64, aarch64, ppc64le & loongaarch64"
+        case [("AIX" | "OS400"), _]:
+            valid = True
```
[the remark this answers is not in the document]
Comment (h-vetinari) [reply tests/test_dependencies.py]: I've turned this into "x86", "x86_64", "s390x" now, which is at least somewhat plausible, together with a TODO to verify the arches.

File: setup.cfg (hunk 5 above)
Comment (h-vetinari) [reply setup.cfg]: > The rule is: every platform is covered by the default requirements, unless there's a reason to bump up the min version.

Done this for linux-ppc & win-arm64 now

File: setup.cfg (hunk 6 above)
Comment (h-vetinari) [reply setup.cfg]: I doubt we go back to Python 3.7 & numpy 1.14.3, but I've just kept the defaults for win-arm64 now.

**Comment by h-vetinari** [comment]
> Thanks @h-vetinari. This looks quite good overall.

I switched this now to use the default pins for linux-ppc & win-arm64. However, I'm still proposing to _not_ distinguish PyPy from CPython for language level >=3.9. Otherwise I think I've responded to all review feedback now.

**Comment by rgommers** [comment]
Cut a release: https://pypi.org/project/oldest-supported-numpy/2023.8.3/#files
